packages/compiler/test/server/completion.ts

```diff
@@ -0,0 +1,245 @@
+import { createTestServerHost } from "./test-server-host.js";
+
+describe("compiler: server: completion", () => {
+  it("completes globals", async () => {
+    const completions = await complete(
+      `
+      model M {
+        s: ┆
+      }
+      `
+    );
+
+    check(completions, [
+      {
+        label: "int32",
+        insertText: "int32",
+        kind: CompletionItemKind.Keyword,
+        documentation: undefined,
+      },
+      {
+        label: "Map",
+        insertText: "Map",
+        kind: CompletionItemKind.Class,
+        documentation: undefined,
+      },
+    ]);
+  });
+
+  it("completes decorators", async () => {
+    const completions = await complete(
+      `
+    @┆
+    model M {}
+    `
+    );
+
+    check(completions, [
+      {
+        label: "doc",
+        insertText: "doc",
+        kind: CompletionItemKind.Function,
+        documentation: undefined,
+      },
+    ]);
+  });
+
+  it("completes partial identifiers", async () => {
+    const completions = await complete(
+      `
+      model M {
+        s: stri┆
+      }
+      `
+    );
+
+    check(completions, [
+      {
+        label: "string",
+        insertText: "string",
+        kind: CompletionItemKind.Keyword,
+        documentation: undefined,
+      },
+    ]);
+  });
+
+  it("completes partial identifier with astral character", async () => {
+    const completions = await complete(
+      `
+      model 𐌰𐌲𐌰𐌲𐌰𐌲 {}
+      model M {
+        s: 𐌰𐌲┆
+      }
+      `
+    );
+
+    check(completions, [
+      {
+        label: "𐌰𐌲𐌰𐌲𐌰𐌲",
+        insertText: "𐌰𐌲𐌰𐌲𐌰𐌲",
+        kind: CompletionItemKind.Class,
+        documentation: undefined,
+      },
+    ]);
+  });
+
+  it("completes namespace members", async () => {
+    const completions = await complete(
+      `
+      namespace N {
+        model A {}
+        model B {}
+      }
+
+      model M extends N.┆
+      `
+    );
+
+    check(
+      completions,
+      [
+        {
+          label: "A",
+          insertText: "A",
+          kind: CompletionItemKind.Class,
+          documentation: undefined,
+        },
+        {
+          label: "B",
+          insertText: "B",
+          kind: CompletionItemKind.Class,
+          documentation: undefined,
+        },
+      ],
+      {
+        allowAdditionalCompletions: false,
+      }
+    );
+  });
+
+  it("completes template parameter uses", async () => {
+    const completions = await complete(
+      `
+      model Template<Param> {
+        prop: ┆
+      }
+      `
+    );
+
+    check(completions, [
+      {
+        label: "Param",
+        insertText: "Param",
+        kind: CompletionItemKind.Struct,
+        documentation: undefined,
+      },
+    ]);
+  });
+
+  it("completes sibling in namespace", async () => {
+    const completions = await complete(
+      `
+      namespace N {
+        model A {}
+        model B extends ┆
```

Comment:
Does this one work fine? I remember it wasn't able to figure out before if the cursor was in the extends section or inside the model body so couldn't complete other model as it would confuse when trying to write properties

Reply:
It does work:

![image](https://user-images.githubusercontent.com/75470/151416920-40e8f768-67e5-4ba6-a1f6-86238083c429.png)

We don't need to know that we're in the extends section for this test, just that we're in a type reference. And I think we can actually find out we're in the heritage here if we needed to for some reason.  But I think the problem you're thinking about would be offering `extends` as a completion before the user types it or `{`? That is harder, still doesn't work, and not tested here.

Reply:
oh right that must have been the issue